### PR TITLE
New covergen tool

### PR DESCRIPTION
**Pull request by geek1011**
cc @shermp @NiLuJe

### CODE REVIEW

**Comment by geek1011** [comment]
TODO:
- [ ] Additional filter options
- [ ] Actually test it
- [ ] Some refactoring
- [ ] Merge geek1011/koboutils#1
- [ ] Documentation

**Comment by geek1011** [comment]
I've tested this, and it works well on my Mini and Aura2v1. For epubs, it now shows the actual cover instead of a screenshot of the first page with a black border. For kepubs, it now consistently shows the cover image, even for books without the cover-image property or the meta tag. There also isn't an extra black border around the covers anymore either, and the cover fills the space properly. I've checked it on the view details, the home screen, the sleep screen, the library list, and the library grid.

To regenerate the covers from nothing, it seems to be just under twice as fast as the built-in cover generation.

P.S. One slightly unrelated thing I noticed is that nickel seems to like segfaulting if you delete .kobo-images from under it and have PDF books.

This still needs to be tested on Windows to make sure the path handling is correct. I'd also appreciate it if someone with a Libra and a wide variety of books could test this out (use the `--regenerate option`) to make sure it works.

**Comment by geek1011** [comment]
I've added an option to set the aspect ratio of the covers. I personally like 1.5, but 1.6 fills the space on the home screen better (but it looks too stretched to me).

File: covergen/covergen.go [hunk 1]
```diff
@@ -158,15 +168,122 @@ func check(ct kobo.CoverType, kp, iid string) (string, bool, error) {
...
+	if rootfile == "" {
+		return nil, errors.New("could not open ebook: could not find package document")
+	}
+
+	var covers []string
+	for _, f := range zr.File {
+		if strings.TrimLeft(strings.ToLower(f.Name), "/") == strings.TrimLeft(strings.ToLower(rootfile), "/") {
+			doc := etree.NewDocument()
+			if rc, err := f.Open(); err != nil {
+				return nil, fmt.Errorf("could not open container.xml: %w", err)
```
Comment (shermp) [inline covergen/covergen.go]: Looks like a copy paste error here?

Should be something like "could not open package document"?

File: covergen/covergen.go [hunk 2]
```diff
@@ -158,15 +168,122 @@ func check(ct kobo.CoverType, kp, iid string) (string, bool, error) {
...
+			} else if _, err = doc.ReadFrom(rc); err != nil {
+				rc.Close()
+				return nil, fmt.Errorf("could not parse container.xml: %w", err)
```
Comment (shermp) [inline covergen/covergen.go]: Ditto

**Comment by shermp** [comment]
Do you have any plans to support PNG and/or SVG (input) cover images?

**Comment by geek1011** [comment]
Yes, those are supported right now (in terms of the format itself), but for the color encoding, I'll need to add conversion so rez can use it.

**Comment by shermp** [comment]
Ah yes. See that now.

Was looking at the resize code, which only deals with YCbCr colourspace.

**Comment by shermp** [comment]
BTW: I'm happy for you to re-license any of the image code from KU if you wish. By this point, it's mostly all your code anyway :p

Main reason for using AGPL3 in KU is because of FBInk. And FBInk has been re-licensed to plain GPL now, so I can probably change KU.

**Comment by geek1011** [comment]
> Was looking at the resize code, which only deals with YCbCr colourspace.

I was considering switching resize libraries, as all are pretty fast on x86, but then I realized some of my scripts run kepubify and seriesmeta on the Kobo itself. I also don't want to pull in too many dependencies.

> BTW: I'm happy for you to re-license any of the image code from KU if you wish. By this point, it's mostly all your code anyway :p

Yes, at first I did pull it from KU, but I realized since I was re-doing the research again, I might as well write it from scratch.

**Comment by geek1011** [comment]
I just tested this while setting up my Clara, and it worked perfectly and had a huge performance benefit to looking through the library. I'm going to merge it and make a release once I merge the koboutils PR.

**Pull request merged**